packages/go_router/lib/src/redirection.dart

```diff
@@ -2,94 +2,182 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/foundation.dart';
+
 import 'configuration.dart';
 import 'logging.dart';
 import 'match.dart';
 import 'matching.dart';
 
 /// A GoRouter redirector function.
-// TODO(johnpryan): make redirector async
-// See https://github.com/flutter/flutter/issues/105808
-typedef RouteRedirector = RouteMatchList Function(RouteMatchList matches,
-    RouteConfiguration configuration, RouteMatcher matcher,
-    {Object? extra});
+typedef RouteRedirector = Future<RouteMatchList> Function(
+    BuildContext, Future<RouteMatchList>, RouteConfiguration, RouteMatcher,
+    {List<RouteMatchList>? redirectHistory, Object? extra});
 
 /// Processes redirects by returning a new [RouteMatchList] representing the new
 /// location.
-RouteMatchList redirect(RouteMatchList prevMatchList,
-    RouteConfiguration configuration, RouteMatcher matcher,
-    {Object? extra}) {
-  // Store each redirect to detect loops
-  final List<RouteMatchList> redirects = <RouteMatchList>[prevMatchList];
-
-  // Keep looping until redirecting is done
-  while (true) {
-    final RouteMatchList currentMatches = redirects.last;
+Future<RouteMatchList> redirect(
```

Comment:
What do you think of making all these `FutureOr`? Benefits: 

1. Makes this breaking change a bit easier for folks migrating to v5
2. Folks don't need to do anything special for the common synchronous scenario

Drawback: our code may be a bit more complicated.

This seems like a good trade-off given folks are unlikely to use `SynchronousFuture` when they should.

Reply:
It is also against flutter style guide https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-futureort

developer can also mark a synchronous method a async if they don't want to bother with SynchronousFuture, and it is not too bad, it will be introduce a one frame delay before page transition. I don't think this is worth it to breaking flutter style guide in this case.